### PR TITLE
draw: Fix circle anchors, remove above/below, fix measure

### LINKED ISSUE #252: Circles' top/bottom anchors are incorrect
`top` and `bottom` are swapped.

MWE:

```typst
#import "@preview/cetz:0.1.2"

#set page(width: auto, height: auto, margin: 1cm)

#cetz.canvas(length: 0.5cm, {
    import cetz.draw: *

    circle((0, 0), name: "bob", radius: 5, fill: black, stroke: none)

    let anchors = (
        "top",      "top-left",     "top-right",
        "bottom",   "bottom-left",  "bottom-right",
        "center",   "left",         "right",
    )

    for pos in anchors {
        content(
            "bob." + pos,
            padding: 0.35,
            frame: "rect",
            fill: rgb(255, 255, 255, 100%, stroke: 1pt + black),
            text(size: 0.75em, pos)
        )
    }
})
```

From 12 o'clock, clockwise: 'bottom', 'top-right', 'right', 'bottom-right', 'top', 'bottom-left', 'left', 'top-left', and back to 'bottom'.

Looks to be from here, top and bottom should have `+/-` `ry` swapped:

https://github.com/johannes-wolf/cetz/blob/282e400b4e0ff583f2a3cabbffa507b94b6f53db/src/draw.typ#L552-L569

### LINKED ISSUE #253: `above`, `below`, and other "default" anchors are not affected by rotations
I think it makes sense to put this in a separate issue.

When working on my MWE for #252, I noticed something peculiar. While the circle's `top` and `bottom` anchors are swapped, `above` and `below` seem to work fine:

<details>
<summary>Code</summary>

```typst
#import "@preview/cetz:0.1.2"

#set page(width: auto, height: auto, margin: 1cm)

#cetz.canvas(length: 1cm, {
    import cetz.draw: *

    circle((0, 0), name: "bob", radius: 5, stroke: 1pt + black)
    line((0, 0), (0, 3.5), mark: (end: ">", size: 0.25, fill: red), stroke: 1pt + red)
    line((0, 0), (3.5, 0), mark: (end: ">", size: 0.25, fill: green), stroke: 1pt + green)
    content((0, 3.75), $x$)
    content((3.75, 0), $y$)

    let anchors = (
        "top",      "top-left",     "top-right",
        "bottom",   "bottom-left",  "bottom-right",
        "center",   "left",         "right",
        "above",
        "below",
    )

    for pos in anchors {
        circle("bob." + pos, radius: 0.2, fill: gray, stroke: 0.5pt + black)

        let (y, anchor) = if pos == "above" or pos == "below" {
            (+0.3, "bottom")
        } else {
            (-0.3, "top")
        }

        content(
            (rel: (0, y)), anchor: anchor,
            frame: "rect", fill: white, stroke: 1pt + gray, padding: 0.15,
            text(size: 0.85em, raw(pos))
        )
    }
})
```

</details>

This led me to snoop around and find where (I'm pretty sure) `above` and `below` are set, and I found something else.

https://github.com/johannes-wolf/cetz/blob/282e400b4e0ff583f2a3cabbffa507b94b6f53db/src/canvas.typ#L121-L154

Looking at this code, it seems to me that:

1.  Elements' default anchors are set based on their AABB.
2.  Then, `above` and `below` are derived from the default `top` and `bottom` anchors.
3.  Then, the `custom-anchors-ctx` function for the element gets run. In the case of `circle`, the `top` and `bottom` anchors are overwritten with the (currently incorrect) ones defined in [`_circle-anchors`](https://github.com/johannes-wolf/cetz/blob/282e400b4e0ff583f2a3cabbffa507b94b6f53db/src/draw.typ#L552-L569).
4.  Since `_circle-anchors` does not define `above`/`below`, they remain as they were, derived from the AABB.
5.  Then, the new `custom-anchors-ctx` anchors are transformed according to the current context before being inserted into the `anchors` list.

This results in `above` and `below` being based on the element's AABB, but `top` and `bottom` being based on the current transformation context:

<details>
<summary>Code</summary>

```typst
#cetz.canvas(length: 1cm, {
    import cetz.draw: *

    rotate(+22.5deg)
    {
        circle((0, 0), name: "bob", radius: 5, stroke: 1pt + black)
        line((0, 0), (0, 3.5), mark: (end: ">", size: 0.25, fill: red), stroke: 1pt + red)
        line((0, 0), (3.5, 0), mark: (end: ">", size: 0.25, fill: green), stroke: 1pt + green)
        content((0, 3.75), $x$)
        content((3.75, 0), $y$)
    }
    rotate(-22.5deg)

    let anchors = (
        "top",      "top-left",     "top-right",
        "bottom",   "bottom-left",  "bottom-right",
        "center",   "left",         "right",
        "above",
        "below",
    )

    for pos in anchors {
        circle("bob." + pos, radius: 0.2, fill: gray, stroke: 0.5pt + black)

        let (y, anchor) = if pos == "above" or pos == "below" {
            (+0.3, "bottom")
        } else {
            (-0.3, "top")
        }

        // https://github.com/johannes-wolf/cetz/issues/252 -- fixed manually to demonstrate
        // different issue
        let pos = if pos == "top" { "bottom" } else if pos == "bottom" { "top" } else { pos }
        content(
            (rel: (0, y)), anchor: anchor,
            frame: "rect", fill: white, stroke: 1pt + gray, padding: 0.15,
            text(size: 0.85em, raw(pos))
        )
    }
})
```

</details>

In this case, it manifests as `above` and `below`; but if there were any other cases of an element returning only some anchors from `custom-anchors-ctx`, then **any** anchors not returned would remain in their default from-AABB state. The fact that most anchors _do_ seem to get properly transformed and the comment saying, "add alternate names", both lead me to believe they are meant to be affected by rotations.

### PR DESCRIPTION
**Pull request by johannes-wolf**
Fixes #252, #253 and #247 

- Fixes the measure function (already fixed in the rework)
- Removes above/below anchors
- Fixes type checking for type(style.stroke) == stroke for marks
- Fixes circle and arc anchors
- Miter calculation was off for lines != 1pt in width.

This fixes some more or less serious bugs, so I would like to release a new version after merge.

### CODE REVIEW

**Comment by matthew-e-brown** [comment]
> Removes above/below anchors

Found some stowaways in `draw.content` :P

https://github.com/johannes-wolf/cetz/blob/b918e084ecfa1ff318001384e02a2dd78360c7e2/src/draw.typ#L862-L876

File: src/draw.typ [hunk 1]
```diff
@@ -339,26 +332,37 @@
...
-      let miter = (calc.tan(90deg - style.angle / 2) +
-                   calc.sin(style.angle / 2)) * width / 2
-      if 2 * miter / width <= limit {
+      let angle = calc.abs(style.angle)
+      let miter = if angle == 180deg {
+        width / 2
+      } else if angle == 0deg or angle == 360deg {
+        0
+      } else {
+        (1 / calc.sin(angle / 2) * width / 2)
```
Comment (johannes-wolf) [inline src/draw.typ:362]: Tbh. I do not know what I was thinking when implementing the previous one…

File: src/draw.typ [hunk 2]
```diff
@@ -339,26 +332,37 @@
...
-      (style.stroke.at("thickness", default: 1pt),
-       style.stroke.at("miter-limit", default: 4),
-       style.stroke.at("join", default: "miter"))
-    } else if type(style.stroke) == stroke {
-      (style.stroke.thickness,
-       style.stroke.miter-limit,
-       style.stroke.join)
+    // This is the only way to "cast" to a stroke type in typst <= 8.0
+    let stroke = line(stroke: style.stroke).stroke
+    let (width, limit, join) = if type(stroke) == typst-stroke {
```
Comment (fenjalien) [inline src/draw.typ:337]: This shouldn't need an if statement from the above line.

File: tests/marks/ref.png
Comment (fenjalien) [inline tests/marks/ref.png:1]: Should this test be moved under `test/mark/`?

File: tests/circle/anchors/test.typ [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  })
+}))
+
+#box(stroke: 2pt + red, canvas({
+  import draw: *
+  
+  rotate(45deg)
+  circle((0,0), radius: (2,3), name: "c")
+  for-each-anchor("c", name => {
+    content("c."+name, box(fill: white)[#name])
```
Comment (fenjalien) [inline tests/circle/anchors/test.typ:19]: I think the bounds are wrong for the bottom-left and top-right labels, they shouldn't be overlapping the edges...

File: tests/rect/anchors/test.typ
Comment (fenjalien) [inline tests/rect/anchors/test.typ:1]: Same issue as circle/anchors

File: tests/viewport/ref.png
Comment (fenjalien) [inline tests/viewport/ref.png:1]: whats caused the padding change? and why does it not affect the rotated C?

**Review by fenjalien — COMMENTED** [review]
Whats the motivation behind removing `above` and `below`?

File: tests/circle/anchors/test.typ (hunk 3 above)
Comment (johannes-wolf) [reply tests/circle/anchors/test.typ:19]: Oh, yes. There is a problem with bounds of `content` getting rotated when it should not.

**Comment by johannes-wolf** [comment]
> Whats the motivation behind removing `above` and `below`?

I think it is not a good idea to have the same anchor twice. If we have above/below they should be different from top/bottom, e.g., in having a padding to the latter.

**Comment by johannes-wolf** [comment]
Leaving this opened until all addressed bugs got fixed in 0.2.0. This PR won't get merged!

**Comment by johannes-wolf** [comment]
Closed in favor of #314.

**Pull request closed without merging**